addons/hr_work_entry_contract/models/hr_contract.py

```diff
@@ -51,37 +51,55 @@ def _get_interval_leave_work_entry_type(self, interval, leaves, bypassing_codes)
+        employees_by_calendar = defaultdict(lambda: self.env['hr.employee'])
+        for contract in self:
+            employees_by_calendar[contract.resource_calendar_id] |= contract.employee_id
+
+        all_resources = self.employee_id.resource_id
+
+        attendances_by_calendar = {
+            calendar: calendar._attendance_intervals_batch(
+                start_dt,
+                end_dt,
+                resources=employees.resource_id,
+                tz=pytz.timezone(calendar.tz)
+            ) for calendar, employees in employees_by_calendar.items()}
+
+        resource_calendar_leaves = self.env['resource.calendar.leaves'].search([
+            ('time_type', '=', 'leave'),
+            # ('calendar_id', '=', self.id), --> Get all the time offs
+            ('resource_id', 'in', [False] + all_resources.ids),
+            ('date_from', '<=', datetime_to_string(end_dt)),
+            ('date_to', '>=', datetime_to_string(start_dt)),
+            ('company_id', '=', self.env.company.id),
+        ])
+        # {resource: resource_calendar_leaves}
+        leaves_by_resource = defaultdict(lambda: self.env['resource.calendar.leaves'])
+        for leave in resource_calendar_leaves:
+            leaves_by_resource[leave.resource_id.id] |= leave
+
         for contract in self:
             employee = contract.employee_id
             calendar = contract.resource_calendar_id
             resource = employee.resource_id
             tz = pytz.timezone(calendar.tz)
-            start_dt = pytz.utc.localize(date_start) if not date_start.tzinfo else date_start
-            end_dt = pytz.utc.localize(date_stop) if not date_stop.tzinfo else date_stop
 
-            attendances = calendar._attendance_intervals_batch(
-                start_dt, end_dt, resources=resource, tz=tz
-            )[resource.id]
+            attendances = attendances_by_calendar[calendar][resource.id]
 
             # Other calendars: In case the employee has declared time off in another calendar
             # Example: Take a time off, then a credit time.
             # YTI TODO: This mimics the behavior of _leave_intervals_batch, while waiting to be cleaned
             # in master.
             resources_list = [self.env['resource.resource'], resource]
-            resource_ids = [False, resource.id]
-            leave_domain = [
-                ('time_type', '=', 'leave'),
-                # ('calendar_id', '=', self.id), --> Get all the time offs
-                ('resource_id', 'in', resource_ids),
-                ('date_from', '<=', datetime_to_string(end_dt)),
-                ('date_to', '>=', datetime_to_string(start_dt)),
-                ('company_id', '=', self.env.company.id),
-            ]
             result = defaultdict(lambda: [])
             tz_dates = {}
-            for leave in self.env['resource.calendar.leaves'].search(leave_domain):
+            for leave in leaves_by_resource[False] | leaves_by_resource[resource.id]:
```

Comment:
We actually do not need to do an union of leaves
```suggestion
            for leave in itertools.chain(leaves_by_resource[False], leaves_by_resource[resource.id]):
```